garden_ai/client.py

```diff
@@ -87,27 +92,56 @@ def __init__(
-            GroupsClient.scopes.resource_server
-        )
-        self.search_authorizer = self._create_authorizer(
-            SearchClient.scopes.resource_server
-        )
-        self.compute_authorizer = self._create_authorizer(COMPUTE_RESOURCE_SERVER_NAME)
-        self.search_client = (
-            SearchClient(authorizer=self.search_authorizer)
-            if not search_client
-            else search_client
-        )
-        self.garden_authorizer = self._create_authorizer(
-            GardenClient.scopes.resource_server
-        )
+        # If auth_client is type AuthClient or None, do an
+        # Authorization Code Grant and make RefreshTokenAuthorizers.
+        # If auth_client is type ConfidentialAppAuthClient, do a
+        # Client Credentials Grant and make ClientCredentialsAuthorizers
+        if (
+            isinstance(auth_client, AuthClient)
+            and not isinstance(auth_client, ConfidentialAppAuthClient)
+        ) or not auth_client:
+            self.auth_client = (
+                NativeAppAuthClient(self.client_id) if not auth_client else auth_client
+            )
+            self.openid_authorizer = self._create_authorizer(
+                AuthClient.scopes.resource_server
+            )
+            self.groups_authorizer = self._create_authorizer(
+                GroupsClient.scopes.resource_server
+            )
+            self.search_authorizer = self._create_authorizer(
+                SearchClient.scopes.resource_server
+            )
+            self.compute_authorizer = self._create_authorizer(
+                COMPUTE_RESOURCE_SERVER_NAME
+            )
+            self.search_client = (
+                SearchClient(authorizer=self.search_authorizer)
+                if not search_client
+                else search_client
+            )
+            self.garden_authorizer = self._create_authorizer(
+                GardenClient.scopes.resource_server
+            )
+        elif isinstance(auth_client, ConfidentialAppAuthClient):
+            self.auth_client = auth_client
+            self.openid_authorizer = ClientCredentialsAuthorizer(
+                self.auth_client,
+                f"{AuthClient.scopes.openid} {AuthClient.scopes.email}",
+            )
+            self.groups_authorizer = ClientCredentialsAuthorizer(
+                self.auth_client, GroupsClient.scopes.view_my_groups_and_memberships
+            )
+            self.search_authorizer = ClientCredentialsAuthorizer(
+                self.auth_client, SearchClient.scopes.all
+            )
+            self.compute_authorizer = ClientCredentialsAuthorizer(
+                self.auth_client, Client.FUNCX_SCOPE
+            )
+            self.search_client = SearchClient(authorizer=self.search_authorizer)
+            self.garden_authorizer = ClientCredentialsAuthorizer(
+                self.auth_client, GardenClient.scopes.test_scope
+            )
```

Comment:
Currently the `_create_authorizer` method is responsible for stashing the user's email in the local datastore after the call to Globus Auth goes through. It seems like this isn't breaking anything yet, but I think you'll want to find a way to make sure that some email address is populated in the database after the ConfidentialAppAuthClient login.